src/cargo/sources/overlay.rs

```diff
@@ -0,0 +1,141 @@
+use std::task::ready;
+
+use crate::sources::IndexSummary;
+
+use super::source::{MaybePackage, Source};
+
+/// A `Source` that overlays a one source over another, pretending that the packages
+/// available in the overlay are actually available in the other one.
+///
+/// This is a massive footgun and a terrible idea, so we do not (and never will)
+/// expose this publicly. However, it is useful for some very specific private
+/// things, like locally verifying a bunch of packages at a time before any of
+/// them have been published.
+pub struct OverlayConfusionAttack<'gctx> {
```

Comment:
Frankly, I think `OverlaySource` is a better fitting name 🙂

Reply:
That name definitely makes more sense, yes. But in office hours someone suggested giving it a scary-looking name to make people less likely to misuse it...

Reply:
I like `OverlaySource` being the suffix.  Maybe `DependencyConfusionThreatOverlaySource`?